org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/format/LSPFormatter.java

```diff
@@ -25,94 +23,77 @@
-import org.eclipse.lsp4e.LanguageServiceAccessor;
-import org.eclipse.lsp4e.LanguageServiceAccessor.LSPDocumentInfo;
+import org.eclipse.lsp4e.LanguageServers;
+import org.eclipse.lsp4e.LanguageServers.LanguageServerDocumentExecutor;
+import org.eclipse.lsp4e.VersionedEdits;
 import org.eclipse.lsp4j.DocumentFormattingParams;
 import org.eclipse.lsp4j.DocumentRangeFormattingParams;
 import org.eclipse.lsp4j.FormattingOptions;
 import org.eclipse.lsp4j.Position;
 import org.eclipse.lsp4j.Range;
 import org.eclipse.lsp4j.ServerCapabilities;
-import org.eclipse.lsp4j.TextEdit;
+import org.eclipse.lsp4j.TextDocumentIdentifier;
 import org.eclipse.ui.editors.text.EditorsUI;
 import org.eclipse.ui.texteditor.AbstractDecoratedTextEditorPreferenceConstants;
 
 public class LSPFormatter {
-
-	public record VersionedFormatRequest(int version, CompletableFuture<List<? extends TextEdit>> edits) {
-		public VersionedFormatRequest() {
-			this(-1, CompletableFuture.completedFuture(Collections.emptyList()));
+	public CompletableFuture<Optional<VersionedEdits>> requestFormatting(@NonNull IDocument document, @NonNull ITextSelection textSelection) throws BadLocationException {
+		URI uri = LSPEclipseUtils.toUri(document);
+		if (uri == null) {
+			return CompletableFuture.completedFuture(Optional.empty());
 		}
-	}
+		LanguageServerDocumentExecutor executor = LanguageServers.forDocument(document).withFilter(LSPFormatter::supportsFormatting);
+		FormattingOptions formatOptions = getFormatOptions();
+		TextDocumentIdentifier docId = new TextDocumentIdentifier(uri.toString());
 
-	private @NonNull LSPDocumentInfo getLSPDocumentInfo(Collection<@NonNull LSPDocumentInfo> infos) {
-		// TODO consider a better strategy for that, maybe iterate on all LS until one
-		// gives a result
-		return infos.iterator().next();
-	}
+		DocumentRangeFormattingParams rangeParams = getRangeFormattingParams(document, textSelection, formatOptions,
+				docId);
+
+		DocumentFormattingParams params = getFullFormatParams(formatOptions, docId);
 
-	public void applyEdits(@NonNull IDocument document, List<? extends TextEdit> edits, int version)
-			throws ConcurrentModificationException {
-		Collection<@NonNull LSPDocumentInfo> infos = LanguageServiceAccessor.getLSPDocumentInfosFor(document,
-				LSPFormatter::supportsFormatting);
-		if (getLSPDocumentInfo(infos).getVersion() != version) {
-			throw new ConcurrentModificationException();
-		} else {
-			try {
-				LSPEclipseUtils.applyEdits(document, edits);
-			} catch (BadLocationException e) {
-				LanguageServerPlugin.logError(e);
+		// TODO: Could refine this algorithm: at present this grabs the first non-null response but the most functional
+		// implementation (if a text selection is present) would try all the servers in turn to see if they supported
+		// range formatting, falling back to a full format if unavailable
+		return executor.computeFirst((w, ls) -> {
+			final ServerCapabilities capabilities = w.getServerCapabilities();
```

Comment:
I am not sure how to improve the code, but I do not like that we need to evaluate the capabilities twice, and prepare the parameters for document and range formatting in advance.
Would it be better to use first one executor with capability filter as in the if, call computeFirst for rangeFormatting on that executor, and only if the list is empty then prepare a second executor for documentFormatting?

What do you think, is that doable?

Reply:
Would you mind if we did that as a future refinement? This implementation mirrors the existing behaviour. I need to get this initial port done as my boss wants me to investigate the resource exhaustion issue I raised yesterday, as it's having customer impact.

Reply:
Evaluating the capabilities and creating the parameters is going to be very cheap. It's more of an issue that if the first LS to respond only supports full formatting then it will beat another server that could have done a more sophisticated format - so for that reason your change makes sense (I had given it some thought already). But at least the first to respond is a bit less arbitrary than just the first one in the list, as the old code did!

In practice, I doubt there will really be multiple LS offering formatting for a document. As we discussed before with Mickael, it's more likely that there will be a primary LS providing most of the brains, and *possibly* one or two others offering things like SonarCube integration.

Reply:
> Would you mind if we did that as a future refinement? This implementation mirrors the existing behaviour. I need to get this initial port done as my boss wants me to investigate the resource exhaustion issue I raised yesterday, as it's having customer impact.

Yes, that is fine for me, this is good enough, so let us merge it, and let us improve it. I can have a look in the meanwhile at the semantic highlight adaptation.